pkg/costattribution/tracker.go

```diff
@@ -0,0 +1,371 @@
+	return tracker
+}
+
+func (t *Tracker) hasSameLabels(labels []string) bool {
+	return slices.Equal(t.labels, labels)
+}
+
+var bufferPool = sync.Pool{
+	New: func() interface{} {
+		return new(bytes.Buffer)
+	},
+}
+
+func (t *Tracker) cleanupTrackerAttribution(key string) {
+	t.observedMtx.Lock()
+	defer t.observedMtx.Unlock()
+	delete(t.observed, key)
+}
+
+func (t *Tracker) IncrementActiveSeries(lbs labels.Labels, now time.Time) {
+	if t == nil {
+		return
+	}
+	t.updateCounters(lbs, now.Unix(), 1, 0, 0, nil, true)
+}
+
+func (t *Tracker) DecrementActiveSeries(lbs labels.Labels) {
+	if t == nil {
+		return
+	}
+	t.updateCounters(lbs, -1, -1, 0, 0, nil, false)
+}
+
+func (t *Tracker) Collect(out chan<- prometheus.Metric) {
+	switch t.state {
+	case Overflow:
+		out <- prometheus.MustNewConstMetric(t.activeSeriesPerUserAttribution, prometheus.GaugeValue, t.overflowCounter.activeSerie.Load(), t.overflowLabels[:len(t.overflowLabels)-1]...)
+		out <- prometheus.MustNewConstMetric(t.receivedSamplesAttribution, prometheus.CounterValue, t.overflowCounter.receivedSample.Load(), t.overflowLabels[:len(t.overflowLabels)-1]...)
+		out <- prometheus.MustNewConstMetric(t.discardedSampleAttribution, prometheus.CounterValue, t.overflowCounter.totalDiscarded.Load(), t.overflowLabels...)
+	case Normal:
+		// Collect metrics for all observed keys
+		t.observedMtx.RLock()
+		defer t.observedMtx.RUnlock()
+		for key, o := range t.observed {
+			if key == "" {
+				continue
+			}
+			keys := strings.Split(key, string(sep))
+
+			keys = append(keys, t.userID)
+			if o.activeSerie.Load() > 0 {
+				out <- prometheus.MustNewConstMetric(t.activeSeriesPerUserAttribution, prometheus.GaugeValue, o.activeSerie.Load(), keys...)
+			}
+			if o.receivedSample.Load() > 0 {
+				out <- prometheus.MustNewConstMetric(t.receivedSamplesAttribution, prometheus.CounterValue, o.receivedSample.Load(), keys...)
+			}
+			o.discardedSampleMtx.Lock()
+			for reason, discarded := range o.discardedSample {
+				out <- prometheus.MustNewConstMetric(t.discardedSampleAttribution, prometheus.CounterValue, discarded.Load(), append(keys, reason)...)
+			}
+			o.discardedSampleMtx.Unlock()
+		}
+	}
+	if t.totalFailedActiveSeries.Load() > 0 {
+		out <- prometheus.MustNewConstMetric(t.failedActiveSeriesDecrement, prometheus.CounterValue, t.totalFailedActiveSeries.Load(), t.userID)
+	}
+}
+
+func (t *Tracker) IncrementDiscardedSamples(lbs []mimirpb.LabelAdapter, value float64, reason string, now time.Time) {
+	if t == nil {
+		return
+	}
+	t.updateCountersWithLabelAdapter(lbs, now.Unix(), 0, 0, value, &reason, true)
+}
+
+func (t *Tracker) IncrementReceivedSamples(lbs []mimirpb.LabelAdapter, value float64, now time.Time) {
+	if t == nil {
+		return
+	}
+	t.updateCountersWithLabelAdapter(lbs, now.Unix(), 0, value, 0, nil, true)
+}
+
+func (t *Tracker) IncrementActiveSeriesFailure() {
+	if t == nil {
+		return
+	}
+	t.totalFailedActiveSeries.Add(1)
+}
```

Comment:
I'm still not convinced this belongs here.

What did fail here? `costattribution.Tracker` didn't fail. It's the caller (`activeseries.ActiveSeries`) who failed to retrieve labels for a series it should know.
So it's the caller who should register the failure.

`costattribution.Tracker` wasn't even called here, so it can't fail.

Reply:
when activeseries.ActiveSeries failed to get index, it is calling this method to increment failure, would you prefer ingester do this job? the advantage of having it here is it goes to cost attribution prometheus registry instead of ops registry.

Reply:
When `activeseries.ActiveSeries` fails to do something, it should increment its failed metric, not someone else's (this component's), because this component didn't fail.

Reply:
make ingester work in commit 1b0fb005cc0ce0c144f2ca83e21e428c8e70aa03